### PR TITLE
refactor(general): avoid async method inside `forEach` (WIP)

### PR DESCRIPTION
**Pull request by Hweinstock**
## Problem
usage of an async function inside a `forEach`, which is almost always not desired. `forEach` will not await the results leading to race conditions and inconsistent orderings.

## Solution
- add lint rule
- migrate existing cases
---

- Treat all work as PUBLIC. Private `feature/x` branches will not be squash-merged at release time.
- Your code changes must meet the guidelines in [CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/eslint-plugin-aws-toolkits/test/rules/no-inline-async-foreach.test.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        'list.forEach(asyncFunctionOrNot)',
+        'list.forEach(() => console.log(x))',
+        'list.forEach(function () {})',
+    ],
+
+    invalid: [
+        { code: 'list.forEach(async (a) => await Promise.resolve(a * a))', errors: [errMsg] },
+        { code: 'list.forEach(async (a: any) => console.log(x))', errors: [errMsg] },
+        { code: 'list.forEach((a) => a.forEach(async (b) => a * b))', errors: [errMsg] },
+        { code: 'list.forEach(async function () {})', errors: [errMsg] },
```
Comment (hayemaxi) [inline plugins/eslint-plugin-aws-toolkits/test/rules/no-inline-async-foreach.test.ts:22]: what about
```
const fn = async (item) => {}
list.forEach(fn)
```

Comment (Hweinstock) [reply plugins/eslint-plugin-aws-toolkits/test/rules/no-inline-async-foreach.test.ts]: See recent commits, also added a few more edge cases.

File: .eslintrc.js [hunk 2]
```diff
@@ -178,6 +178,7 @@ module.exports = {
         'aws-toolkits/no-console-log': 'error',
         'aws-toolkits/no-json-stringify-in-log': 'error',
         'aws-toolkits/no-printf-mismatch': 'error',
+        'aws-toolkits/no-async-in-foreach': 'error',
```
Comment (justinmk3) [inline .eslintrc.js:181]: should we just ban `forEach`? it is kind of pointless.

Comment (Hweinstock) [reply .eslintrc.js:181]: Yea, I think banning it out right has a few advantages. Additionally, determining if a function is async is definitely non-trivial. This work is moved to https://github.com/aws/aws-toolkit-vscode/pull/6281
